### PR TITLE
Autoscheduler bugfix: inlining trivial Funcs into extern Funcs.

### PR DESCRIPTION
**Pull request by SanderVocke**
The autoscheduler can fail if it encounters a "trivial func" (func for which inlining is equal or cheaper than the lowest possible memory cost) which is consumed by an extern func. It will try to inline the trivial Func into the extern Func, which is illegal.

This PR adds a test exposing the problem, and consecutively a fix.

### CODE REVIEW

**Comment by SanderVocke** [comment]
Looks like the Travis CI testing script does not run "make test_auto_schedule". Is that intentional? If it did, in the first commit of this PR (before the bugfix), I expect the build to fail on the "extern.cpp" test due to the bug described.

**Comment by SanderVocke** [comment]
Also looks like the first commit failed a build due to exceeding the time limit.

File: src/AutoSchedule.cpp [hunk 1]
```diff
@@ -3138,9 +3138,17 @@ bool inline_all_trivial_functions(const vector<Function> &outputs,
             for (int j = i + 1; j < (int)order.size() - (int)outputs.size(); ++j) {
                 internal_assert(order[i] != order[j]);
                 Function f2 = env.at(order[j]);
-                debug(5) << "Inline trivial function \"" << f1.name()
-                         << "\" inside \"" << f2.name() << "\"\n";
-                inline_function(f2, f1);
+
+                if( f2.has_extern_definition() &&  !f1.is_wrapper() ) {
```
Comment (steven-johnson) [inline src/AutoSchedule.cpp]: Halide style is `if (cond)` rather than `if( cond )`

File: test/auto_schedule/extern.cpp [hunk 2]
```diff
@@ -70,6 +72,46 @@ int main(int argc, char **argv) {
...
+// should not attempt to inline into the extern consumer.
+void test_case_2() {
+
+    ImageParam input(UInt(8), 2);
+    Var x("x"), y("y");
+    Func f0("f0"), f1("f1"), f2("f2"), g("g");
+
+    f0(x, y) = input(x, y) * 2;
+
+    // make f1, which is not a wrapper, but is trivial to inline
```
Comment (psuriana) [inline test/auto_schedule/extern.cpp]: Upper-case: Make

File: test/auto_schedule/extern.cpp [hunk 3]
```diff
@@ -33,7 +33,9 @@ extern "C" DLLEXPORT int translate(buffer_t *in, int dx, int dy, buffer_t *out)
 
 using namespace Halide;
 
-int main(int argc, char **argv) {
+// Test a pipe with several extern-defined Funcs.
+void test_case_1() {
+
```
Comment (psuriana) [inline test/auto_schedule/extern.cpp]: Extra line

File: test/auto_schedule/extern.cpp [hunk 4]
```diff
@@ -70,6 +72,46 @@ int main(int argc, char **argv) {
...
+    p.auto_schedule(target);
+
+    // Inspect the schedule
+    g.print_loop_nest();
+
+}
+
+int main(int argc, char **argv) {
+
```
Comment (psuriana) [inline test/auto_schedule/extern.cpp]: Extra line

File: test/auto_schedule/extern.cpp [hunk 5]
```diff
@@ -70,6 +72,46 @@ int main(int argc, char **argv) {
     // Inspect the schedule
     g.print_loop_nest();
 
+}
+
+// Test with an extern Func which consumes a trivial Func; autoscheduler
+// should not attempt to inline into the extern consumer.
+void test_case_2() {
+
```
Comment (psuriana) [inline test/auto_schedule/extern.cpp]: Extra line

File: test/auto_schedule/extern.cpp [hunk 6]
```diff
@@ -70,6 +72,46 @@ int main(int argc, char **argv) {
...
+    ImageParam input(UInt(8), 2);
+    Var x("x"), y("y");
+    Func f0("f0"), f1("f1"), f2("f2"), g("g");
+
+    f0(x, y) = input(x, y) * 2;
+
+    // make f1, which is not a wrapper, but is trivial to inline
+    // into the next extern Func (because print() has no cost)
+    f1(x, y) = print( f0(x, y) );
```
Comment (psuriana) [inline test/auto_schedule/extern.cpp]: `print(f0(x, y))`. No extra spaces

File: test/auto_schedule/extern.cpp (hunk 2 above)
Comment (psuriana) [reply test/auto_schedule/extern.cpp]: Also maybe *Create* instead of *Make*

File: test/auto_schedule/extern.cpp [hunk 7]
```diff
@@ -70,6 +72,46 @@ int main(int argc, char **argv) {
     // Inspect the schedule
     g.print_loop_nest();
```
Comment (psuriana) [inline test/auto_schedule/extern.cpp]: Extra line

File: test/auto_schedule/extern.cpp [hunk 8]
```diff
@@ -70,6 +72,46 @@ int main(int argc, char **argv) {
...
+
+    // Auto-schedule the pipeline
+    Target target = get_jit_target_from_environment();
+    Pipeline p(g);
+
+    p.auto_schedule(target);
+
+    // Inspect the schedule
+    g.print_loop_nest();
+
```
Comment (psuriana) [inline test/auto_schedule/extern.cpp]: Extra line

**Comment by SanderVocke** [comment]
Thanks, processed your reviews.

Is Halide in the Google C++ style?

**Comment by abadams** [comment]
It's similar. We use four-space indentation. There's a .clang-format file
in the root of the repo which tries to encode the style.

**Comment by steven-johnson** [comment]
> There's a .clang-format file in the root of the repo which tries to encode the style

Note that it's close but not perfect; IIRC there were some idioms we couldn't quite encode.

**Pull request merged**